src/IO/S3Common.cpp

```diff
@@ -98,104 +102,320 @@ ServerSideEncryptionKMSConfig getSSEKMSConfig(const std::string & config_elem, c
-    if (!from.secret_access_key.empty())
-        secret_access_key = from.secret_access_key;
-    if (!from.session_token.empty())
-        session_token = from.session_token;
+RequestSettings::RequestSettings(const DB::Settings & settings, bool validate_settings)
+{
+    updateFromSettings(settings, /* if_changed */false, validate_settings);
+    finishInit(settings, validate_settings);
+}
 
-    if (!from.headers.empty())
-        headers = from.headers;
-    if (!from.region.empty())
-        region = from.region;
+void RequestSettings::updateFromSettings(
+    const DB::Settings & settings, bool if_changed, bool validate_settings)
+{
+    for (auto & field : allMutable())
+    {
+        const auto setting_name = "s3_" + field.getName();
+        if (settings.has(setting_name) && (!if_changed || settings.isChanged(setting_name)))
+        {
+            set(field.getName(), settings.get(setting_name));
+        }
+    }
 
-    server_side_encryption_customer_key_base64 = from.server_side_encryption_customer_key_base64;
-    server_side_encryption_kms_config = from.server_side_encryption_kms_config;
+    normalizeSettings();
+    if (validate_settings)
+        validateUploadSettings();
+}
 
-    if (from.use_environment_credentials.has_value())
-        use_environment_credentials = from.use_environment_credentials;
+void RequestSettings::updateIfChanged(const RequestSettings & settings)
+{
+    for (auto & setting : settings.all())
+    {
+        if (setting.isValueChanged())
+            set(setting.getName(), setting.getValue());
+    }
+}
 
-    if (from.use_insecure_imds_request.has_value())
-        use_insecure_imds_request = from.use_insecure_imds_request;
+void RequestSettings::normalizeSettings()
+{
+    if (!storage_class_name.value.empty() && storage_class_name.changed)
+        storage_class_name = Poco::toUpperInPlace(storage_class_name.value);
+}
 
-    if (from.expiration_window_seconds.has_value())
-        expiration_window_seconds = from.expiration_window_seconds;
+void RequestSettings::finishInit(const DB::Settings & settings, bool validate_settings)
+{
+    normalizeSettings();
+    if (validate_settings)
+        validateUploadSettings();
+
+    /// NOTE: it would be better to reuse old throttlers
+    /// to avoid losing token bucket state on every config reload,
+    /// which could lead to exceeding limit for short time.
+    /// But it is good enough unless very high `burst` values are used.
+    if (UInt64 max_get_rps = isChanged("max_get_rps") ? get("max_get_rps").get<UInt64>() : settings.s3_max_get_rps)
+    {
+        size_t default_max_get_burst = settings.s3_max_get_burst
+            ? settings.s3_max_get_burst
+            : (Throttler::default_burst_seconds * max_get_rps);
 
-    if (from.no_sign_request.has_value())
-        no_sign_request = from.no_sign_request;
+        size_t max_get_burst = isChanged("max_get_burts") ? get("max_get_burst").get<UInt64>() : default_max_get_burst;
```

Comment:
Noticed the typos while refactoring this (`max_get_burts` -> `max_get_burst`, `max_put_burts` -> `max_put_burst`).

Aside from the typos, I don't quite see how this is supposed to work. Who could populate RequestSettings::max_get_burts considering the setting is not declared inside this class?